File: src/services/fhir-integration.service.ts
import { Injectable, inject } from '@angular/core';

@Injectable({
  providedIn: 'root'
})
export class FhirIntegrationService {

  // EPIC / SMART ON FHIR CONFIGURATION
  // In a real production app, the Client ID comes from Epic App Orchard
  private readonly CLIENT_ID = 'pocketgull-sandbox-client-id'; 
  
  // Scopes define what data we ask the patient for permission to read
  private readonly SCOPES = 'launch/patient openid fhirUser patient/Observation.read patient/Condition.read patient/DiagnosticReport.read patient/DocumentReference.read patient/MedicationRequest.read';

  private readonly EPIC_AUTH_URL = 'https://fhir.epic.com/interconnect-fhir-oauth/oauth2/authorize';
  private readonly EPIC_TOKEN_URL = 'https://fhir.epic.com/interconnect-fhir-oauth/oauth2/token';

  /**
   * Step 1: Redirects the user to the Epic/MyChart login page to grant access
   */
  authorize() {
    // Determine the absolute redirect URI dynamically based on the current window location
    const redirectUri = window.location.origin + '/fhir-callback';
    
    // Create standard OAuth 2.0 authorization query string
    const params = new URLSearchParams({
      response_type: 'code',
      client_id: this.CLIENT_ID,
      redirect_uri: redirectUri,
      scope: this.SCOPES,
      state: crypto.randomUUID(), // Prevent CSRF attacks
      aud: 'https://fhir.epic.com/interconnect-fhir-oauth/api/FHIR/R4' // Target FHIR Server URL
    });

    const authorizeUrl = `${this.EPIC_AUTH_URL}?${params.toString()}`;
    
    // Redirect browser to Epic
    window.location.href = authorizeUrl;
  }

  /**
   * Step 2: Exchanges the authorization code for an Access Token
   * This is typically called by the /fhir-callback route component.
   */
  async handleCallback(code: string): Promise<boolean> {
    try {
      const redirectUri = window.location.origin + '/fhir-callback';
      
      // In a real application without a backend proxy, the Epic OAuth token endpoint 
      // sometimes requires PKCE rather than a raw POST if using a pure SPA. 
      // Assuming a standard sandbox flow for demonstration:
      const params = new URLSearchParams({
        grant_type: 'authorization_code',
        code: code,
        redirect_uri: redirectUri,
        client_id: this.CLIENT_ID
      });

      const response = await fetch(this.EPIC_TOKEN_URL, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: params.toString()
      });

      if (!response.ok) {
        throw new Error(`Token exchange failed: ${response.statusText}`);
      }

      const tokenData = await response.json();
      
      // Store token securely (in memory or encrypted local storage)
      sessionStorage.setItem('epic_access_token', tokenData.access_token);
      sessionStorage.setItem('epic_patient_id', tokenData.patient); // The patient context id

      console.log('Successfully connected to Epic MyChart!');
      return true;

    } catch (e) {
      console.error('[FhirIntegrationService] Auth callback error:', e);
      return false;
    }
  }

  /**
   * Step 3: Use the Token to pull down patient data.
   * Note: We don't implement the full REST pulls here to avoid bloat, 
   * but this method proves the connection works.
   */
  async fetchPatientProfile() {
      const token = sessionStorage.getItem('epic_access_token');
      const patientId = sessionStorage.getItem('epic_patient_id');

      if (!token || !patientId) return null;

      try {
        const response = await fetch(`https://fhir.epic.com/interconnect-fhir-oauth/api/FHIR/R4/Patient/${patientId}`, {
            headers: {
                'Authorization': `Bearer ${token}`,
                'Accept': 'application/json'
            }
        });

        if (response.ok) {
            return await response.json();
        }
      } catch (e) {
          console.error('[FhirIntegrationService] Search error:', e);
      }
      return null;
  }

  public hasValidToken(): boolean {
    if (typeof window === 'undefined') return false;
    return !!sessionStorage.getItem('epic_access_token');
  }
}

File: pocketgull_flutter/lib/services/fhir_integration_service.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

/// FHIR / SMART-on-FHIR integration service.
///
/// Flutter parity with Angular `fhir-integration.service.ts`.
/// Handles OAuth 2.0 authorization flow with Epic MyChart and
/// token management for FHIR R4 API access.

class FhirIntegrationService {
  /// Epic SMART-on-FHIR configuration.
  static const String _clientId = 'pocketgull-sandbox-client-id';
  static const String _scopes =
      'launch/patient openid fhirUser '
      'patient/Observation.read patient/Condition.read '
      'patient/DiagnosticReport.read patient/DocumentReference.read '
      'patient/MedicationRequest.read';
  static const String _epicAuthUrl =
      'https://fhir.epic.com/interconnect-fhir-oauth/oauth2/authorize';
  static const String _epicTokenUrl =
      'https://fhir.epic.com/interconnect-fhir-oauth/oauth2/token';

  String? _accessToken;
  String? _patientId;

  /// Whether we have a valid access token.
  bool get isConnected => _accessToken != null;
  String? get patientId => _patientId;

  /// Build the OAuth 2.0 authorization URL for the given redirect URI.
  ///
  /// On mobile this would be opened via `url_launcher` or a custom tab.
  Uri buildAuthorizeUrl(String redirectUri) {
    return Uri.parse(_epicAuthUrl).replace(queryParameters: {
      'response_type': 'code',
      'client_id': _clientId,
      'redirect_uri': redirectUri,
      'scope': _scopes,
      'state': DateTime.now().millisecondsSinceEpoch.toString(),
      'aud': 'https://fhir.epic.com/interconnect-fhir-oauth/api/FHIR/R6',
    });
  }

  /// Exchange an authorization code for an access token.
  Future<bool> handleCallback(String code, String redirectUri) async {
    try {
      final response = await http.post(
        Uri.parse(_epicTokenUrl),
        headers: {'Content-Type': 'application/x-www-form-urlencoded'},
        body: {
          'grant_type': 'authorization_code',
          'code': code,
          'redirect_uri': redirectUri,
          'client_id': _clientId,
        },
      );

      if (response.statusCode != 200) {
        debugPrint('[FHIR] Token exchange failed: ${response.statusCode}');
        return false;
      }

      final tokenData = jsonDecode(response.body);
      _accessToken = tokenData['access_token'] as String?;
      _patientId = tokenData['patient'] as String?;

      debugPrint('[FHIR] Successfully connected to Epic MyChart!');
      return true;
    } catch (e) {
      debugPrint('[FHIR] Token exchange error: $e');
      return false;
    }
  }

  /// Fetch a FHIR resource from the connected server.
  Future<Map<String, dynamic>?> fetchResource(String resourceType, {String? id}) async {
    if (_accessToken == null) return null;
    try {
      final url = id != null
          ? 'https://fhir.epic.com/interconnect-fhir-oauth/api/FHIR/R6/$resourceType/$id'
          : 'https://fhir.epic.com/interconnect-fhir-oauth/api/FHIR/R6/$resourceType?patient=$_patientId';
      final response = await http.get(
        Uri.parse(url),
        headers: {
          'Authorization': 'Bearer $_accessToken',
          'Accept': 'application/fhir+json',
        },
      );
      if (response.statusCode == 200) {
        return jsonDecode(response.body) as Map<String, dynamic>;
      }
    } catch (e) {
      debugPrint('[FHIR] Fetch error: $e');
    }
    return null;
  }

  /// Disconnect from the FHIR server.
  void disconnect() {
    _accessToken = null;
    _patientId = null;
  }
}

final fhirIntegrationProvider = Provider<FhirIntegrationService>((ref) {
  return FhirIntegrationService();
});
